x-pack/plugins/reporting/server/lib/store/store.ts

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ElasticsearchServiceSetup } from 'src/core/server';
+import { LevelLogger } from '../';
+import { ReportingCore } from '../../';
+import { LayoutInstance } from '../../export_types/common/layouts';
+import { indexTimestamp } from './index_timestamp';
+import { mapping } from './mapping';
+import { Report } from './report';
+
+export const statuses = {
+  JOB_STATUS_PENDING: 'pending',
+  JOB_STATUS_PROCESSING: 'processing',
+  JOB_STATUS_COMPLETED: 'completed',
+  JOB_STATUS_WARNINGS: 'completed_with_warnings',
+  JOB_STATUS_FAILED: 'failed',
+  JOB_STATUS_CANCELLED: 'cancelled',
+};
+
+interface AddReportOpts {
+  timeout: number;
+  created_by: string | boolean;
+  browser_type: string;
+  max_attempts: number;
+}
+
+interface UpdateQuery {
+  index: string;
+  id: string;
+  if_seq_no: unknown;
+  if_primary_term: unknown;
+  body: { doc: Partial<Report> };
+}
+
+/*
+ * A class to give an interface to historical reports in the reporting.index
+ * - track the state: pending, processing, completed, etc
+ * - handle updates and deletes to the reporting document
+ * - interface for downloading the report
+ */
+export class ReportingStore {
+  public readonly indexPrefix: string;
+  public readonly indexInterval: string;
+
+  private client: ElasticsearchServiceSetup['legacy']['client'];
+  private logger: LevelLogger;
+
+  constructor(reporting: ReportingCore, logger: LevelLogger) {
+    const config = reporting.getConfig();
+    const elasticsearch = reporting.getElasticsearchService();
+
+    this.client = elasticsearch.legacy.client;
+    this.indexPrefix = config.get('index');
+    this.indexInterval = config.get('queue', 'indexInterval');
+
+    this.logger = logger;
+  }
+
+  private async createIndex(indexName: string) {
+    return this.client
+      .callAsInternalUser('indices.exists', {
+        index: indexName,
+      })
+      .then((exists) => {
+        if (exists) {
+          return exists;
+        }
+
+        const indexSettings = {
+          number_of_shards: 1,
+          auto_expand_replicas: '0-1',
+        };
+        const body = {
+          settings: indexSettings,
+          mappings: {
+            properties: mapping,
+          },
+        };
+
+        return this.client
+          .callAsInternalUser('indices.create', {
+            index: indexName,
+            body,
+          })
+          .then(() => true)
+          .catch((err: Error) => {
+            const isIndexExistsError = err.message.match(/resource_already_exists_exception/);
+            if (isIndexExistsError) {
+              // Do not fail a job if the job runner hits the race condition.
+              this.logger.warn(`Automatic index creation failed: index already exists: ${err}`);
```

Comment:
This fixes a bug that was a FIXME comment: https://github.com/elastic/kibana/blob/2f67cbf/x-pack/plugins/reporting/server/lib/esqueue/helpers/create_index.js#L93

Continuing from index creation failure, when the index already exists, is now actually logged